_posts/2022-11-29-graphdl/graphdl.Rmd

```diff
@@ -0,0 +1,547 @@
+        print(inputs)
+        print("in update, x is")
+        print(x)
+        return (inputs + x)/2
+
+module = IDoEvolveOverTime()
+out = module(data.x, data.edge_index)
+print("result is:")
+out
+```
+
+    in update, inputs is
+    tensor([[ 6.,  6.],
+            [ 6.,  6.],
+            [29., 29.],
+            [26., 26.],
+            [26., 26.]])
+    in update, x is
+    tensor([[ 1.,  1.],
+            [ 2.,  2.],
+            [ 3.,  3.],
+            [11., 11.],
+            [12., 12.]])
+    result is:
+    tensor([[ 3.5000,  3.5000],
+            [ 4.0000,  4.0000],
+            [16.0000, 16.0000],
+            [18.5000, 18.5000],
+            [19.0000, 19.0000]])
+
+In `update()`, we have access to both the final message aggregate (`inputs`) and the nodes' prior states (`x`). Here, I'm just averaging those two.
+
+At this point, we've successfully acquainted ourselves with the three stages of message passing: acting on individual messages, aggregating them, and self-updating based on past state and new information. But none of our models so far could be called a neural network, since there was no learning involved.
+
+### Adding parameters
+
+If we look back at the generic message passing formulation:
+
+$$
+\mathbf{x}_i^{(k)} = \gamma^{(k)} \left( \mathbf{x}_i^{(k-1)}, \square_{j \in \mathcal{N}(i)} \, \phi^{(k)}\left(\mathbf{x}_i^{(k-1)}, \mathbf{x}_j^{(k-1)},\mathbf{e}_{j,i}\right) \right)
+$$ we see two places where neural network modules can act on the computation: before message aggregation, and as part of the node update process. First, we illustrate the former option. For example, we can apply a MLP in `forward()`, before the call to `aggregate()`:
+
+```{python}
+class ILearnAndEvolve(MessagePassing):
+    def __init__(self, in_channels, out_channels):
+        super().__init__(aggr = "sum")
+        self.mlp = Seq(Linear(in_channels, out_channels),
+                       ReLU(),
+                       Linear(out_channels, out_channels))
+    def forward(self, x, edge_index):
+        edge_index, _ = add_self_loops(edge_index, num_nodes=x.size(0))
+        x = self.mlp(x)
+        out = self.propagate(edge_index = edge_index, x = x)
+        return out
+    def update(self, inputs, x):
+        return (inputs + x)/2
+
+module = ILearnAndEvolve(2, 2)
+out = module(data.x, data.edge_index)
+print("result is:")
+out
+```
+
+    result is:
+    tensor([[-0.8724, -0.4407],
+            [-0.9056, -0.4623],
+            [-2.0229, -1.1240],
+            [-1.8691, -1.0867],
+            [-1.9024, -1.1082]], grad_fn=<DivBackward0>)
+
+Finally, we can apply network modules in both places, as exemplified next.
+
+### General message passing
+
+We keep the MLP from the previous class, and add a second in `update()`:
+
+```{python}
+class ILearnAndEvolveDoubly(MessagePassing):
+    def __init__(self, in_channels, out_channels):
+        super().__init__(aggr = "sum")
+        self.mlp1 = Seq(Linear(in_channels, out_channels),
```

Comment:
Maybe use a name like `mlp_forward` and `mlp_update`?

Reply:
Makes sense, will do!